subprojects/core/src/main/java/org/gradle/internal/typeconversion/MapNotationConverter.java

```diff
@@ -134,4 +103,68 @@ protected String get(Map<String, Object> args, String key) {
         return str;
     }
 
+    private static class ConvertMethod {
+        private final static Map<Class, ConvertMethod> CONVERT_METHODS = Maps.newHashMap();
```

Comment:
Should be a weak reference, or otherwise allow classes to be collected. This is another example of where we want to cache some inspection of a class.

Also, can you make this non-static and injectable, so we can later make it smarter (eg we might have it, or something it uses, understand that core types don't expire and don't need to be wrapped in a reference, or to receive notifications when a `ClassLoader` has been discarded so that it can release references to types in the `ClassLoader`.